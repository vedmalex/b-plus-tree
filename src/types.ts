import type { ValueType } from './Node'

export type Comparator<K extends ValueType> = (a?: K, b?: K) => number

export type UnaryFunction<T, R> = (source: T) => R

/* eslint:disable:max-line-length */
export function query<T, K extends ValueType>(): UnaryFunction<T, T>
export function query<T, A>(fn1: UnaryFunction<T, A>): UnaryFunction<T, A>
export function query<T, A, B>(
  fn1: UnaryFunction<T, A>,
  fn2: UnaryFunction<A, B>,
): UnaryFunction<T, B>
export function query<T, A, B, C>(
  fn1: UnaryFunction<T, A>,
  fn2: UnaryFunction<A, B>,
  fn3: UnaryFunction<B, C>,
): UnaryFunction<T, C>
export function query<T, A, B, C, D>(
  fn1: UnaryFunction<T, A>,
  fn2: UnaryFunction<A, B>,
  fn3: UnaryFunction<B, C>,
  fn4: UnaryFunction<C, D>,
): UnaryFunction<T, D>
export function query<T, A, B, C, D, E>(
  fn1: UnaryFunction<T, A>,
  fn2: UnaryFunction<A, B>,
  fn3: UnaryFunction<B, C>,
  fn4: UnaryFunction<C, D>,
  fn5: UnaryFunction<D, E>,
): UnaryFunction<T, E>
export function query<T, A, B, C, D, E, F>(
  fn1: UnaryFunction<T, A>,
  fn2: UnaryFunction<A, B>,
  fn3: UnaryFunction<B, C>,
  fn4: UnaryFunction<C, D>,
  fn5: UnaryFunction<D, E>,
  fn6: UnaryFunction<E, F>,
): UnaryFunction<T, F>
export function query<T, A, B, C, D, E, F, G>(
  fn1: UnaryFunction<T, A>,
  fn2: UnaryFunction<A, B>,
  fn3: UnaryFunction<B, C>,
  fn4: UnaryFunction<C, D>,
  fn5: UnaryFunction<D, E>,
  fn6: UnaryFunction<E, F>,
  fn7: UnaryFunction<F, G>,
): UnaryFunction<T, G>
export function query<T, A, B, C, D, E, F, G, H>(
  fn1: UnaryFunction<T, A>,
  fn2: UnaryFunction<A, B>,
  fn3: UnaryFunction<B, C>,
  fn4: UnaryFunction<C, D>,
  fn5: UnaryFunction<D, E>,
  fn6: UnaryFunction<E, F>,
  fn7: UnaryFunction<F, G>,
  fn8: UnaryFunction<G, H>,
): UnaryFunction<T, H>
export function query<T, A, B, C, D, E, F, G, H, I>(
  fn1: UnaryFunction<T, A>,
  fn2: UnaryFunction<A, B>,
  fn3: UnaryFunction<B, C>,
  fn4: UnaryFunction<C, D>,
  fn5: UnaryFunction<D, E>,
  fn6: UnaryFunction<E, F>,
  fn7: UnaryFunction<F, G>,
  fn8: UnaryFunction<G, H>,
  fn9: UnaryFunction<H, I>,
): UnaryFunction<T, I>
export function query<T, A, B, C, D, E, F, G, H, I>(
  fn1: UnaryFunction<T, A>,
  fn2: UnaryFunction<A, B>,
  fn3: UnaryFunction<B, C>,
  fn4: UnaryFunction<C, D>,
  fn5: UnaryFunction<D, E>,
  fn6: UnaryFunction<E, F>,
  fn7: UnaryFunction<F, G>,
  fn8: UnaryFunction<G, H>,
  fn9: UnaryFunction<H, I>,
  ...fns: Array<UnaryFunction<unknown, unknown>>
): UnaryFunction<T, unknown>
/* eslint:enable:max-line-length */

export function query(
  ...fns: Array<UnaryFunction<unknown, unknown>>
): UnaryFunction<unknown, unknown> {
  return queryFromArray(fns)
}

export function identity<T>(x: T): T {
  return x
}

/** @internal */
export function queryFromArray<T, R>(
  fns: Array<UnaryFunction<T, R>>,
): UnaryFunction<T, R> | ((input: T) => UnaryFunction<T, R>) {
  if (fns.length === 0) {
    return identity as UnaryFunction<T, R>
  }

  if (fns.length === 1) {
    return fns[0]
  }

  return (input: T) => {
    let res: T | R = input
    fns.forEach((fn) => {
      res = fn(res as T)
    })
    return res as unknown as R
    // return fns.reduce(
    //   (prev: T, fn: UnaryFunction<T, R>) => fn(prev),
    //   input as any,
    // )
  }
}

// export interface CursorFunction<T, K extends ValueType> {
//   (source: Generator<Cursor<T, K>> | AsyncGenerator<Cursor<T, K>>):
//     | Generator<Cursor<T, K>>
//     | AsyncGenerator<Cursor<T, K>>
// }
