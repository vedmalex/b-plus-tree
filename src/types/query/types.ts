import { Cursor } from '../eval/Cursor'

export interface UnaryFunction<T, R> {
  (source: T): R
}

/* eslint:disable:max-line-length */
export function query<T>(): UnaryFunction<T, T>
export function query<T, A>(fn1: UnaryFunction<T, A>): UnaryFunction<T, A>
export function query<T, A, B>(
  fn1: UnaryFunction<T, A>,
  fn2: UnaryFunction<A, B>,
): UnaryFunction<T, B>
export function query<T, A, B, C>(
  fn1: UnaryFunction<T, A>,
  fn2: UnaryFunction<A, B>,
  fn3: UnaryFunction<B, C>,
): UnaryFunction<T, C>
export function query<T, A, B, C, D>(
  fn1: UnaryFunction<T, A>,
  fn2: UnaryFunction<A, B>,
  fn3: UnaryFunction<B, C>,
  fn4: UnaryFunction<C, D>,
): UnaryFunction<T, D>
export function query<T, A, B, C, D, E>(
  fn1: UnaryFunction<T, A>,
  fn2: UnaryFunction<A, B>,
  fn3: UnaryFunction<B, C>,
  fn4: UnaryFunction<C, D>,
  fn5: UnaryFunction<D, E>,
): UnaryFunction<T, E>
export function query<T, A, B, C, D, E, F>(
  fn1: UnaryFunction<T, A>,
  fn2: UnaryFunction<A, B>,
  fn3: UnaryFunction<B, C>,
  fn4: UnaryFunction<C, D>,
  fn5: UnaryFunction<D, E>,
  fn6: UnaryFunction<E, F>,
): UnaryFunction<T, F>
export function query<T, A, B, C, D, E, F, G>(
  fn1: UnaryFunction<T, A>,
  fn2: UnaryFunction<A, B>,
  fn3: UnaryFunction<B, C>,
  fn4: UnaryFunction<C, D>,
  fn5: UnaryFunction<D, E>,
  fn6: UnaryFunction<E, F>,
  fn7: UnaryFunction<F, G>,
): UnaryFunction<T, G>
export function query<T, A, B, C, D, E, F, G, H>(
  fn1: UnaryFunction<T, A>,
  fn2: UnaryFunction<A, B>,
  fn3: UnaryFunction<B, C>,
  fn4: UnaryFunction<C, D>,
  fn5: UnaryFunction<D, E>,
  fn6: UnaryFunction<E, F>,
  fn7: UnaryFunction<F, G>,
  fn8: UnaryFunction<G, H>,
): UnaryFunction<T, H>
export function query<T, A, B, C, D, E, F, G, H, I>(
  fn1: UnaryFunction<T, A>,
  fn2: UnaryFunction<A, B>,
  fn3: UnaryFunction<B, C>,
  fn4: UnaryFunction<C, D>,
  fn5: UnaryFunction<D, E>,
  fn6: UnaryFunction<E, F>,
  fn7: UnaryFunction<F, G>,
  fn8: UnaryFunction<G, H>,
  fn9: UnaryFunction<H, I>,
): UnaryFunction<T, I>
export function query<T, A, B, C, D, E, F, G, H, I>(
  fn1: UnaryFunction<T, A>,
  fn2: UnaryFunction<A, B>,
  fn3: UnaryFunction<B, C>,
  fn4: UnaryFunction<C, D>,
  fn5: UnaryFunction<D, E>,
  fn6: UnaryFunction<E, F>,
  fn7: UnaryFunction<F, G>,
  fn8: UnaryFunction<G, H>,
  fn9: UnaryFunction<H, I>,
  ...fns: UnaryFunction<any, any>[]
): UnaryFunction<T, {}>
/* eslint:enable:max-line-length */

export function query(
  ...fns: Array<UnaryFunction<any, any>>
): UnaryFunction<any, any> {
  return queryFromArray(fns)
}

export function identity<T>(x: T): T {
  return x
}

/** @internal */
export function queryFromArray<T, R>(
  fns: Array<UnaryFunction<T, R>>,
): UnaryFunction<T, R> {
  if (fns.length === 0) {
    return identity as UnaryFunction<any, any>
  }

  if (fns.length === 1) {
    return fns[0]
  }

  return function queried(input: T): R {
    return fns.reduce(
      (prev: any, fn: UnaryFunction<T, R>) => fn(prev),
      input as any,
    )
  }
}

interface CursorFunction<T> {
  (source: Generator<Cursor<T>> | AsyncGenerator<Cursor<T>>):
    | Generator<Cursor<T>>
    | AsyncGenerator<Cursor<T>>
}
