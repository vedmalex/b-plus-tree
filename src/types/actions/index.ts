import { remove } from './remove'
import { update } from './update'

export { remove, update }
