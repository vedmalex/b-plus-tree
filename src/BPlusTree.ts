import { Node } from './Node'
import type { PortableBPlusTree } from './PortableBPlusTree'
import type { Cursor } from './eval/Cursor'
import { remove } from './methods/remove'
import { remove_specific } from './methods/remove_specific'
import { insert } from './methods/insert'
import { count } from './methods/count'
import { size } from './methods/size'
import { find_last_node } from './methods/find_last_node'
import { find_first_key } from './methods/find_first_key'
import { find_first_node } from './methods/find_first_node'
import { find } from './eval/find'
import { list } from './eval/list'
import { sourceIn } from './source/sourceIn'
import { sourceEq } from './source/sourceEq'
import { sourceRange } from './source/sourceRange'
import { sourceEach } from './source/sourceEach'
import { sourceGt } from './source/sourceGt'
import { sourceGte } from './source/sourceGte'
import { sourceLt } from './source/sourceLt'
import { sourceLte } from './source/sourceLte'
import type { ValueType } from './ValueType'
import { sourceEqNulls } from './source/sourceEqNulls'
import { find_first_item } from './methods/find_first_item'
import { find_last_item } from './methods/find_last_item'
import type { PortableNode } from './Node/PortableNode'
import type { Comparator } from './types'
import { compare_keys_primitive } from '../methods/utils/comparator_primitive'
/**
 * tree
 * T - value to be stored
 * K - key
 */
export class BPlusTree<T, K extends ValueType> {
  public t: number // минимальная степень дерева
  public root: number // указатель на корень дерева
  public unique: boolean
  public nodes = new Map<number, Node<T, K>>()
  public comparator: Comparator<K>
  public defaultEmpty: K
  public keySerializer: (keys: Array<K>) => any
  public keyDeserializer: (keys: any) => Array<K>
  protected next_node_id = 0
  get_next_id(): number {
    return this.next_node_id++
  }

  includes(
    keys: Array<K>,
  ): (tree: BPlusTree<T, K>) => Generator<Cursor<T, K>, void> {
    return sourceIn<T, K>(keys)
  }

  equals(key: K): (tree: BPlusTree<T, K>) => Generator<Cursor<T, K>, void> {
    return sourceEq<T, K>(key)
  }

  equalsNulls(
    key: K,
  ): (tree: BPlusTree<T, K>) => Generator<Cursor<T, K>, void> {
    return sourceEqNulls<T, K>(key)
  }

  range(
    from: K,
    to: K,
    fromIncl = true,
    toIncl = true,
  ): (tree: BPlusTree<T, K>) => Generator<Cursor<T, K>, void> {
    return sourceRange<T, K>(from, to, fromIncl, toIncl)
  }

  each(
    forward = true,
  ): (tree: BPlusTree<T, K>) => Generator<Cursor<T, K>, void> {
    return sourceEach<T, K>(forward)
  }

  gt(key: K): (tree: BPlusTree<T, K>) => Generator<Cursor<T, K>, void> {
    return sourceGt<T, K>(key)
  }
  gte(key: K): (tree: BPlusTree<T, K>) => Generator<Cursor<T, K>, void> {
    return sourceGte<T, K>(key)
  }
  lt(key: K): (tree: BPlusTree<T, K>) => Generator<Cursor<T, K>, void> {
    return sourceLt<T, K>(key)
  }
  lte(key: K): (tree: BPlusTree<T, K>) => Generator<Cursor<T, K>, void> {
    return sourceLte<T, K>(key)
  }

  constructor(
    t?: number,
    unique?: boolean,
    comparator?: (a: K, b: K) => number,
    defaultEmpty?: K,
    keySerializer?: (keys: Array<K>) => any,
    keyDeserializer?: (keys: any) => Array<K>,
  ) {
    this.t = t ?? 32
    this.unique = unique ?? false
    this.root = Node.createLeaf(this).id
    this.comparator = comparator ?? compare_keys_primitive
    this.defaultEmpty =
      defaultEmpty ?? (Number.NEGATIVE_INFINITY as unknown as K)
    this.keySerializer = keySerializer ?? ((keys: Array<K>) => keys)
    this.keyDeserializer = keyDeserializer ?? ((keys: any) => keys as Array<K>)
  }

  static serialize<T, K extends ValueType>(
    tree: BPlusTree<T, K>,
  ): PortableBPlusTree<T, K> {
    const { t, root, unique, nodes, next_node_id } = tree
    return {
      t,
      next_node_id,
      root,
      unique,
      nodes: [...nodes.values()].map((n) => Node.serialize(n)),
    }
  }
  static createFrom<T, K extends ValueType>(
    stored: PortableBPlusTree<T, K>,
  ): BPlusTree<T, K> {
    const res = new BPlusTree<T, K>()
    BPlusTree.deserialize(res, stored)
    return res
  }
  static deserialize<T, K extends ValueType>(
    tree: BPlusTree<T, K>,
    stored: PortableBPlusTree<T, K>,
  ): void {
    const { t, next_node_id, root, unique, nodes } = stored
    if (t) {
      tree.nodes.clear()
      tree.t = t
      tree.next_node_id = next_node_id
      tree.root = root
      tree.unique = unique
      nodes.forEach((n) => {
        const node = Node.deserialize<T, K>(n, tree)
        tree.nodes.set(n.id, node)
      })
    } else {
      // key pair serialiation
      for (const [key, value] of Object.entries(stored)) {
        tree.insert(key as K, value as unknown as T)
      }
    }
  }

  find(
    key?: K,
    {
      skip = 0,
      take = -1,
      forward = true,
    }: { skip?: number; take?: number; forward?: boolean } = {},
  ): Array<T> {
    return find(this, key, { skip, take, forward })
  }

  list({
    skip = 0,
    take = -1,
    forward = true,
  }: { skip?: number; take?: number; forward?: boolean } = {}): Array<T> {
    return list(this, { skip, take, forward })
  }

  findFirst(key: K): T {
    const node = find_first_node(this, key)
    const index = find_first_item(node.keys, key, this.comparator)
    return node.pointers[index]
  }

  findLast(key: K): T {
    const node = find_last_node(this, key)
    const index = find_last_item(node.keys, key, this.comparator)
    return node.pointers[index]
  }

  cursor(key: K): Cursor<T, K> {
    const node = find_last_node(this, key)
    const index = find_first_key(node.keys, key, this.comparator)
    const value = node.pointers[index]
    return { node: node.id, pos: index, key, value, done: value === undefined }
  }

  reset(): void {
    this.next_node_id = 0
    this.nodes.clear()
    this.root = Node.createLeaf(this).id
  }

  get min(): K {
    return this.nodes.get(this.root).min
  }
  get max(): K {
    return this.nodes.get(this.root).max
  }
  get size(): number {
    return size(this.nodes.get(this.root))
  }
  node(id: number): Node<T, K> {
    return this.nodes.get(id)
  }
  count(key: K): number {
    if (key == undefined) key = null
    return count(key, this.nodes.get(this.root))
  }
  insert(key: K, value: T): boolean {
    if (key == null) key = this.defaultEmpty
    return insert(this, key, value)
  }

  removeSpecific(key: K, specific: (pointers: T) => boolean): Array<[K, T]> {
    //TODO: допилить эту штуку
    if (key == undefined) key = null
    return remove_specific(this, key, specific)
  }
  remove(key: K): Array<[K, T]> {
    if (key == undefined) key = null
    return remove(this, key, false)
  }

  removeMany(key: K): Array<[K, T]> {
    if (key == undefined) key = null
    return remove(this, key, true)
  }

  toJSON(): {
    t: number
    unique: boolean
    root: PortableNode<T, K> & { errors: Array<string> }
  } {
    return {
      t: this.t,
      unique: this.unique,
      root: this.nodes.get(this.root).toJSON(),
    }
  }
}
