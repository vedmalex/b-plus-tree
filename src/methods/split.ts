import { BPlusTree } from '../types/BPlusTree'
import { Node } from '../types/Node'

export function split(this: BPlusTree, node: Node) {
  let new_node = new Node() //Создаем новый узел

  // Перенаправляем right и left указатели
  new_node.right = node.right
  if (node.right) node.right.left = new_node
  node.right = new_node
  new_node.left = node

  // Перемещаем t - 1 значений и соответствующих им указателей в new_node
  const mid_key = node.keys[this.t]
  new_node.key_num = this.t - 1
  node.key_num = this.t
  for (let i = 0; i <= new_node.key_num - 1; i++) {
    new_node.keys[i] = node.keys[i + this.t + 1]
    // if (node.leaf) {
    new_node.pointers[i] = node.pointers[i + this.t + 1]
    // } else {
    new_node.children[i] = node.children[i + this.t + 1]
    // }
  }

  // похоже это если узел не является leaf ???
  new_node.children[new_node.key_num] = node.children[2 * this.t]

  if (node.leaf) {
    ++new_node.key_num
    new_node.leaf = true
    // Перемещаем в new_node оставшийся при разбиении элемент mid_key в начало
    new_node.keys.unshift(node.keys[this.t])
    new_node.pointers.unshift(node.pointers[this.t])
  }

  if (node == this.root) {
    this.root = new Node() // Создаем новый корень T.root
    this.root.keys[0] = mid_key
    this.root.children[0] = node
    this.root.children[1] = new_node
    this.root.key_num = 1
    node.parent = this.root
    new_node.parent = this.root
  } else {
    new_node.parent = node.parent
    const parent = node.parent

    // Ищем позицию mid_key в отце
    let pos = 0
    while (pos < parent.key_num && parent.keys[pos] < mid_key) {
      ++pos
    }

    // Добавляем mid_key в отца и направляем ссылку из него на new_node
    for (let i = parent.key_num; i >= pos + 1; i--) {
      parent.keys[i] = parent.keys[i - 1]
    }
    for (let i = parent.key_num + 1; i >= pos + 2; i--) {
      parent.children[i] = parent.children[i - 1]
    }
    parent.keys[pos] = mid_key
    parent.children[pos + 1] = new_node
    ++parent.key_num

    if (parent.key_num == 2 * this.t) {
      split.call(this, parent)
    }
  }
  // фиксируем размеры массива освобождая память
  new_node.commit()
  node.commit()
}
