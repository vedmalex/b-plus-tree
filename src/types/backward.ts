import { Node } from './Node'
import { ValueType } from '../btree'
import { BPlusTree } from './BPlusTree'
import { getItems } from './getItems'

function get_items_from_array({
  array,
  skip = 0,
  take = -1,
  forward = true,
}: {
  array: Array<any>
  skip?: number
  take?: number
  forward?: boolean
}) {
  const result = []
  if (take == -1) take = array.length - skip
  if (forward) {
    const start = skip
    const end = skip + take
    for (let i = start; i < end; i++) result.push(array[i])
  } else {
    const length = array.length
    const start = length - skip - 1
    const end = start - take
    for (let i = start; i > end; i--) result.push(array[i])
  }
  return result
}

// сравнить reverse() и обход с конца и с начала
export function walkthrought({
  tree,
  key = undefined,
  skip = 0,
  take = -1,
  forward = false,
}: {
  tree: BPlusTree
  key?: ValueType
  skip?: number
  take?: number
  forward?: boolean
}) {
  let node: Node
  if (key === undefined) {
    // если ключ поиска не задан
    node = forward
      ? // ищем первый элемент
        tree.find_first_node(tree.min())
      : // или последний
        tree.find_last_node(tree.max())
  } else {
    // ищем заданный элемент
    node = forward ? tree.find_first_node(key) : tree.find_last_node(key)
  }
  let result = []
  let toBeSkipped = skip
  let taken = take
  do {
    const lres = getItems(node, key)
    // const lres = getData(node)
    if (lres.length > 0) {
      const resLen = lres.length
      let lskip = toBeSkipped
      if (toBeSkipped > 0) {
        lskip -= resLen
      }
      if (lskip <= 0) {
        if (taken == -1) {
          // или берем все или только часть вырезаем
          if (toBeSkipped == 0) {
            result.push(...get_items_from_array({ forward, array: lres }))
          } else {
            // вырезаем от начального элемента
            result.push(
              ...get_items_from_array({
                forward,
                array: lres,
                skip: toBeSkipped,
              }),
            )
            // следующий слок читаем сначала
            toBeSkipped = 0
          }
        } else {
          if (toBeSkipped + taken > 0 && toBeSkipped + taken < resLen) {
            result.push(
              ...get_items_from_array({
                forward,
                array: lres,
                skip: toBeSkipped,
                take: taken,
              }),
            )
            taken = 0
            break
          } else if (toBeSkipped + taken > 0 && toBeSkipped + taken >= resLen) {
            if (toBeSkipped == 0) {
              result.push(...get_items_from_array({ forward, array: lres }))
            } else {
              result.push(
                ...get_items_from_array({
                  forward,
                  array: lres,
                  skip: toBeSkipped,
                }),
              )
            }
            taken -= resLen - toBeSkipped
            toBeSkipped = 0
          } else {
            // skipped + taken == 0
            // мы полуили все что нужно
            // и тут оказались случайно
            break
          }
        }
      } else {
        toBeSkipped = lskip
      }
      node = forward ? node.right : node.left
    } else {
      // первый элемент не найден в дереве
      break
    }
  } while (node != null)
  return result
}
