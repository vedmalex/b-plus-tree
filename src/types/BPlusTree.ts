import print from 'print-tree'
import { Node } from './Node'
import { ValueType } from '../btree'
import { remove } from '../methods/remove'
import { insert } from '../methods/insert'
import { find_key } from '../methods/find_key'
import { updateValue } from '../methods/updateValue'

export class BPlusTree {
  public t: number // минимальная степень дерева
  public root: Node // указатель на корень дерева
  public unique: boolean
  constructor(t: number, unique: boolean) {
    this.root = Node.createLeaf()
    this.t = t
    this.unique = unique
  }
  find(key: ValueType): ReturnType<typeof find_key> {
    return find_key.call(this, key)
  }
  insert(key: ValueType, value: any) {
    return insert.call(this, key, value)
  }
  remove(key: ValueType) {
    return remove.call(this, key)
  }
  min() {
    return this.root.min
  }
  max() {
    return this.root.max
  }
  isNodeFull(node: Node) {
    updateValue
    const res =
      (node.leaf ? node.keys.length : node.children.length) > this.t * 2
    node.isFull = res
    return res
  }
  toJSON() {
    return {
      t: this.t,
      unique: this.unique,
      root: this.root.toJSON(),
    }
  }
  print() {
    print(
      this.toJSON().root,
      (node) =>
        `${node.id} <${node.min ?? ''}:${node.max ?? ''}> ${JSON.stringify(
          node.keys,
        )}`,
      (node) => node.children,
    )
  }
}
