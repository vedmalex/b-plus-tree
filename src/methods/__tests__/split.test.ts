import 'jest'
import { split } from '../split'
import { BPlusTree } from '../../types/BPlusTree'

const insertion = [0, 9, 3, 14, 12, 13, 6, 10, 2, 1, 4, 8, 5, 11, 7]

describe('split', () => {
  it('split leaf into two sepate', () => {
    const bpt = new BPlusTree(2, true)
    expect(() =>
      insertion.forEach((i, index) => {
        bpt.insert(i, index)
      }),
    ).not.toThrow()

    expect(() => {
      insertion.forEach((item) => {
        const result = bpt.find(item).keys.indexOf(item)
        if (result == -1) throw new Error(`not found ${item}`)
      })
    }).not.toThrow()

    expect(bpt.toJSON()).toMatchSnapshot('B+Tree')
  })
})
