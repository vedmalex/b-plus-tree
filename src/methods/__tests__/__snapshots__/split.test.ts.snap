// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`split split leaf into two sepate: B+Tree 1`] = `
Object {
  "root": Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "id": 0,
            "key_num": 2,
            "keys": Array [
              0,
              1,
            ],
            "max": 1,
            "min": 0,
            "pointers": Array [
              0,
              9,
            ],
          },
          Object {
            "id": 4,
            "key_num": 2,
            "keys": Array [
              2,
              3,
            ],
            "max": 3,
            "min": 2,
            "pointers": Array [
              8,
              2,
            ],
          },
        ],
        "id": 2,
        "key_num": 1,
        "keys": Array [
          2,
        ],
        "max": 3,
        "min": 0,
      },
      Object {
        "children": Array [
          Object {
            "id": 7,
            "key_num": 2,
            "keys": Array [
              4,
              5,
            ],
            "max": 5,
            "min": 4,
            "pointers": Array [
              10,
              12,
            ],
          },
          Object {
            "id": 8,
            "key_num": 3,
            "keys": Array [
              6,
              7,
              8,
            ],
            "max": 8,
            "min": 6,
            "pointers": Array [
              6,
              14,
              11,
            ],
          },
        ],
        "id": 9,
        "key_num": 1,
        "keys": Array [
          6,
        ],
        "max": 8,
        "min": 4,
      },
      Object {
        "children": Array [
          Object {
            "id": 1,
            "key_num": 3,
            "keys": Array [
              9,
              10,
              11,
            ],
            "max": 11,
            "min": 9,
            "pointers": Array [
              1,
              7,
              13,
            ],
          },
          Object {
            "id": 3,
            "key_num": 3,
            "keys": Array [
              12,
              13,
              14,
            ],
            "max": 14,
            "min": 12,
            "pointers": Array [
              4,
              5,
              3,
            ],
          },
        ],
        "id": 5,
        "key_num": 1,
        "keys": Array [
          12,
        ],
        "max": 14,
        "min": 9,
      },
    ],
    "id": 6,
    "key_num": 2,
    "keys": Array [
      4,
      9,
    ],
    "max": 14,
    "min": 0,
  },
  "t": 2,
  "unique": true,
}
`;
